### PR TITLE
Add Flatpak support

### PR DESCRIPTION
**Pull request by kirbyfan64**
### What exactly is this?

This PR adds a support for building a TSC [Flatpak](https://flatpak.org/). This has a bunch of different benefits, from distribution to contributing.

### Distribution

Flatpak is basically yet another distribution platform for Linux that has been brought up here before. There are a couple of highlights here:

- Supports different repositories, but has an official one, Flathub, meaning we could throw TSC 2.1 up there to make downloading super easy.

### Building

TSC has some hard-to-build dependencies (actually, mostly just CEGUI) that can make contributing difficult. Flatpak-builder largely solves this by having builds in an isolated environment, too.

### IDE support

One other nicety (which is actually what inspired me to do this) is that you get great IDE support. I can open up the CMakeLists.txt in GNOME Builder and immediately have a full development environment with all the dependencies set up. Same apparently goes for KDevelop (FWIW KDE now primarily uses Flatpak for nighties).

### TL;DR

It makes everything easier!

### CODE REVIEW

File: org.secretchronicles.TSC.json [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+            "name": "boost",
+            "buildsystem": "simple",
+            "build-commands": [
+                "./bootstrap.sh --prefix=/app",
+                "./b2 install --with-atomic --with-chrono --with-date_time --with-filesystem --with-thread"
+            ],
+            "sources": [
+                {
+                    "type": "archive",
+                    "url": "https://dl.bintray.com/boostorg/release/1.68.0/source/boost_1_68_0.tar.bz2",
```
Comment (Quintus) [inline org.secretchronicles.TSC.json]: Is this the official Boost URL? If so, leave it so, if not: why not use the canonical project URL?

**Comment by Quintus** [comment]
Overall, looks fine to me. I only had a few minor points, see the respective review comments.

**Comment by Quintus** [comment]
> One other nicety (which is actually what inspired me to do this) is that you get great IDE support. I can open up the CMakeLists.txt in GNOME Builder and immediately have a full development environment with all the dependencies set up. Same apparently goes for KDevelop (FWIW KDE now primarily uses Flatpak for nighties).

Emacs user here :-P

File: org.secretchronicles.TSC.json (hunk 1 above)
Comment (kirbyfan64) [reply org.secretchronicles.TSC.json]: Yeah, it's the official one. As in, if you go to the Boost website and right-click one of the downloads, this is what you get. They've moved from SourceForge to bintray for binary hosting.

**Comment by kirbyfan64** [comment]
> Emacs user here :-P

~~rip~~

FWIW this was inspired by both me moving to Fedora Silverblue (where global packages are discouraged vs containers/Flatpak) and Sydney's troubles in building TSC.

**Comment by xet7** [comment]
@kirbyfan64 

What commands you used to install flatpak and build package? I could not get it working by following flatpak website docs.

**Comment by kirbyfan64** [comment]
@xet7 You're on Debian, right? What exactly wasn't working?

**Comment by xet7** [comment]
@kirbyfan64 

I have wrong syntax. I don't know how to build package.

```
$ flatpak build-init tsc-build org.secretchronicles.TSC 18.08 org.freedesktop.Platform
error: '18.08' is not a valid sdk name: Name can't start with 1
```

**Comment by kirbyfan64** [comment]
@xet7 I think you're mixing up some different stuff... In order to build this, you'd just do:

```
flatpak-builder --force-clean --install MY_BUILD_DIRECTORY-I-usually-use-"build" org.secretchronicles.TSC.json
```

(Note that it won't actually work yet since the CMake support commits haven't been merged, but it will after this PR.) `--install` will install TSC after building.

If you're looking to use it for development, you'll probably want something like GNOME Builder which supports building Flatpaks out-of-the-box. Just open CMakeLists.txt in Builder and watch the magic.

**Comment by Alexander-Wilms** [comment]
Just FYI, when you rebuild the Flatpak several times, using flatpak-builder's --ccache option helps.

**Comment by kirbyfan64** [comment]
@xet7 

- Run `git submodule update --init` to grab the new submodule.

- Your software looks a bit old... GNOME Builder needs to be at least 3.30, or flatpak-builder needs to be a bit newer for `--install` to work. Both of these can be solved through Flatpak (surprise!!)... For a newer Builder, use `flatpak install flathub org.gnome.Builder` and run `org.gnome.Builder`. For a newer flatpak-builder, run `flatpak install flathub org.flatpak.Builder` and run `org.flatpak.Builder` instead of `flatpak-builder`.

- To open CMakeLists.txt, open up GNOME Builder -> click Open on the top -> change the dropdown on the bottom right to *CMake Project*, then open up CMakeLists.txt inside the `src` subdirectory.

**Comment by Alexander-Wilms** [comment]
You need to clone the submodule AS well: https://github.com/Secretchronicles/TSC/blob/f75da33c0a416e1ca036a2041c50faa5a6bca76c/.gitmodules#L24

**Comment by kirbyfan64** [comment]
FWIW as there doesn't seem to really be opposition to this, I'm just going to merge it now a sec, that way the manifest will actually work standalone (yay!)...

**Comment by Quintus** [comment]
Mh, I had [some comments left](https://github.com/Secretchronicles/TSC/pull/639/files) in the patch commit diff. It doesn't look as if you addressed them... I can see that GitHub makes it easy to overlook such comments, so you might still want to take look now?

**Comment by xet7** [comment]
@kirbyfan64 

Please, don't merge PR's yourself. And please do not consider these comments to PR's as "not any real opposition". These comments are not unreal. It's not opposition. I really need to get PR's working myself before merging.

I still get this error:
```
$ flatpak install flathub org.gnome.Builder
error: app/org.gnome.Builder/x86_64/stable needs a later flatpak version (0.10.0)
```
Please do not expect that when it works on your computer that it will work for me. I have spent many hours to get this working, and it still does not work.

**Comment by kirbyfan64** [comment]
> Debian 9

Ohhhh...

Try installing the [version in backports](https://backports.debian.org/Instructions/#index3h2):

    apt-get install -t stretch-backports flatpak flatpak-builder

Now should be able to run:

    flatpak-builder --install --force-clean my-build-directory org.secretchronicles.TSC.json

and it'll all work.

**Comment by xet7** [comment]
I still have errors. I added my current build scripts to repo.
```
$ ./build-flatpak.sh 
Downloading sources
Fetching git repo https://github.com/mattn/mruby-json, ref refs/heads/master
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
Fetching git repo https://github.com/mattn/mruby-md5, ref refs/heads/master
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
Fetching full git repo https://github.com/mattn/mruby-pcre-regexp
Fetching git repo https://github.com/matsumotory/mruby-sleep, ref refs/heads/master
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
Initialized empty Git repository in /home/user/repos/Secretchronicles/TSC/.flatpak-builder/git/_home_user_repos_Secretchronicles_TSC_https_github.com_Secretchronicles_TSC-51VUQZ/
fatal: '/home/user/repos/Secretchronicles/TSC/https:/github.com/Secretchronicles/TSC' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
Failed to download sources: module tsc: Child process exited with code 128
```

**Comment by kirbyfan64** [comment]
@xet7 Ahhhhh sorry that one was all me... I typo'd. :man_facepalming: 

Potential solution: try changing [these lines](https://github.com/Secretchronicles/TSC/blob/devel/org.secretchronicles.TSC.json#L226-L229) to:

```json
{
    "type": "dir",
    "path": "."
}
```

**Comment by xet7** [comment]
@kirbyfan64 

Thanks! Now it installed flatpak and I see TSC on desktop game menu.

How do I make .flatpak package file and add it to be installed from Alexandria?

**Comment by kirbyfan64** [comment]
@xet7 Run:

```
flatpak-builder --repo a-repo-directory my-build-directory org.secretchronicles.TSC.json
flatpak build-export a-repo-directory my-build-directory
```

That being said: the preferred way is to host a Flatpak repo. My ideal, eventual scenario (that I might start work on now) is that CI will automatically build a Flatpak bundle as a build artifact (to make it easy to test different builds). Then, every night, Alexandria will import the latest bundle from the master branch into a nightly repository, which can be easily used by users to install a nightly TSC (you basically have to make a simple text file that users will double-click to add the repo + install it).
